[STALLED] Davies bouldin index

tomron opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->


#### What does this implement/fix? Explain your changes.
Added implementation of Davies Bouldin index (https://en.wikipedia.org/wiki/Davies%E2%80%93Bouldin_index)

Implementation includes - 
* Function code
* Exposing function in the `__init__.py` files
* Testing
* Adding documentation and usage example

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

tommagic committed:
Adding davies_bouldin_index calculation

1. sklearn/metrics/cluster/unsupervised.py - calculation itself
2. sklearn/metrics/cluster/tests/test_unsupervised.py - tests
3. sklearn/metrics/cluster/__init__.py - exposing the function

tomron committed:
Expose function at metrics level

tomron committed:
Adding documentation and usage example

tommagic committed:
Fix - adding davies_bouldin_index to all list

tguillemot commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -255,3 +255,63 @@ def calinski_harabaz_score(X, labels):
+def davies_bouldin_index(X, labels):
+    """Compute the Davies Bouldin index.
+
+    The index is defiend as the ratio of within-cluster
```
defined

tguillemot commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -255,3 +255,63 @@ def calinski_harabaz_score(X, labels):
+    n_labels = len(le.classes_)
+
+    check_number_of_labels(n_labels, n_samples)
+    clusters_data = {}
```
I think it's better to allocate mean_k and d_k with a fixed size and a numpy.array

tomron replied:
Can you please reference me to an example somewhere so I'll modified it accordingly?

jnothman replied:
Please average intracluster distances and centroids in separate numpy arrays.

tomron replied:
done

tguillemot commented:
@tomron Thanks for the PR. I have some questions for you.

I'm not an expert, so maybe the questions are stupid : there are a lot of metrics for clustering in sklearn, so why adding a new one ? 
I've read your doc and indeed the computation is simpler than silhouette but I am not sure that the DBI is something used in practice (maybe I'm wrong). Moreover for a fast computation, we have already the Calinsky Harabaz metric which seems have the same advantages and drawbacks.

Can you give more details and explain the pros/cons in comparaison with Calinsky Harabaz and why we will add DBI ?

Thanks in advance

tommagic committed:
Typo fix: defiend -> defined

tomron commented:
@tguillemot 
Simple answer - I needed it and implemented it so I decided to contribute it.
Comparing to Calinsky Harabaz, DBI is bounded (0-1).
There are some use cases where DBI performs better or more relevant than Calinsky Harabaz or silhouette, or just the relevant measure for the person who uses it.

amueller commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+  >>> from sklearn import metrics
+  >>> from sklearn.metrics import pairwise_distances
+  >>> from sklearn import datasets
+  >>> dataset = datasets.load_iris()
```
you could use ``load_Xy``

amueller commented:
It might be nice to add this to the comparison in #6948. Can you maybe run that and see how the metric fares?

amueller commented:
This has 3543 cites btw.

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+separation between clusters.
+
+For :math:`k` clusters, the Davies–Bouldin index :math:`DB` is given as the
+ratio of within cluster-mean distance to the between means distance.
```
Something here is not right.

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+.. math::
+  DB(k) = \frac{1}{k} \sum_{i=1}^k \max_{i \neq j} D_{ij}
+
+Where :math:`D_ij` is the ratio between the within distances in clusters
```
`ij` -> `{ij}`

tomron replied:
done

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+:math:`i` and :math:`j`.
+
+.. math::
+  D_ij = \frac{\bar{d_i}+\bar{d_j}}{d_ij}
```
`ij` -> `{ij}` x2

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+.. math::
+  D_ij = \frac{\bar{d_i}+\bar{d_j}}{d_ij}
+
+:math:`\bar{d_i}` is the average distance between each point cluster
```
"point cluster" -> "point in"

tomron replied:
done

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+:math:`i` and the centroid of cluster :math:`i`.
+:math:`\bar{d_i}` is the diameter of cluster :math:`i`.
+
+:math:`\bar{d_j}` is the average distance between each point cluster
```
"point cluster" -> "point in"

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+:math:`j` and the centroid of cluster :math:`j`.
+:math:`\bar{d_j}` is the diameter of cluster :math:`j`.
+
+:math:`d_ij` is the Euclidean distance between the centroid of cluster
```
`ij` -> `{ij}`

tomron replied:
done

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+:math:`\bar{d_j}` is the diameter of cluster :math:`j`.
+
+:math:`d_ij` is the Euclidean distance between the centroid of cluster
+:math:`i` and the centroid of cluster :math:`j`.
```
need this to end with `::` for following doctest.

tomron replied:
You mean adding :: in the end of like 1605?

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+:math:`i` and the centroid of cluster :math:`j`.
+
+
+  >>> from sklearn import metrics
```
This should only need to be loaded once in the file. It's already redundantly loaded for two metrics, but this is an error.

tomron replied:
I followed the other examples in the file. It make sense for me that each example should run independent from the others.

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+  >>> from sklearn import datasets
+  >>> dataset = datasets.load_iris()
+  >>> X = dataset.data
+  >>> y = dataset.target
```
This is unused.

tomron replied:
Done

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+  >>> y = dataset.target
+
+In normal usage, the Davies-Bouldin index is applied to the results of a
+cluster analysis.
```
need this to end with `::` for following doctest.

tomron replied:
You mean adding :: in the end of like 1616?

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+
+- The Davies-Bouldin index is generally higher for convex clusters than other
+  concepts of clusters, such as density based clusters like those obtained
+  through DBSCAN.
```
Also: use of centroid distance limits it to Euclidean space.

tomron replied:
done

jnothman commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -146,3 +147,32 @@ def test_calinski_harabaz_score():
+    assert_equal(0., davies_bouldin_index([[-1, -1], [1, 1]] * 10,
+                                          [0] * 10 + [1] * 10))
+
+    # General case (with non numpy arrays)
```
i've not checked the paper, but if the paper has examples you can copy in here, it's best to make the test suite as complete as is reasonable with respect to those examples.

tomron replied:
I didn't find numeric self contained example in the article

jnothman commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -146,3 +147,32 @@ def test_calinski_harabaz_score():
+         [[0, 4], [1, 3]] * 5 + [[3, 1], [4, 0]] * 5)
+    labels = [0] * 10 + [1] * 10 + [2] * 10 + [3] * 10
+    assert_almost_equal(davies_bouldin_index(X, labels),
+                        2*np.sqrt(0.5)/3)
```
Please make sure you test the case where a cluster has a single sample.

tomron replied:
done

jnothman commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -255,3 +255,63 @@ def calinski_harabaz_score(X, labels):
+    for k in range(n_labels):
+        cluster_k = X[labels == k]
+        mean_k = np.mean(cluster_k, axis=0)
+        d_k = np.average(pairwise_distances(cluster_k, mean_k))
```
`mean_k` -> `[mean_k]`

tomron replied:
I improved the code

jnothman commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -255,3 +255,63 @@ def calinski_harabaz_score(X, labels):
+        clusters_data[k] = (mean_k, d_k)
+
+    score = 0
+    for i in range(n_labels):
```
this can be performed with something like:
```
scores = (intra_dists[:, None] + intra_dists) / pairwise_distances(centroids)
# TODO: fix diagonal infs
return np.mean(np.max(scores, axis=1))
```

tomron replied:
I improved the code

tommagic committed:
Test cluster with one sample

Adding a test to validate that the code run correctly when
there is one sample in a cluster (not in all clusters, this is
already validated).

tommagic committed:
Fix mathematic notation

tommagic committed:
Add drawback - euclidean distance only

tommagic committed:
mean_k -> [mean_k]

tommagic committed:
Efficiency and style fixes

1. average intracluster distances and centroids in separate numpy arrays
2. Adjust code for the case where a cluster have single sample

tommagic committed:
Remove unused variable

tommagic committed:
Flakes and pep8 fixes

tguillemot commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+~~~~~~~~~~
+
+- The computation of the Davies-Bouldin index is simpler than the computation
+  of the Silhouette index.
```
Can you add a word that contraty to Calinsky Harabaz, DBI is bounded (0-1) ?

tguillemot commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -255,3 +255,55 @@ def calinski_harabaz_score(X, labels):
+
+    check_number_of_labels(n_labels, n_samples)
+    intra_dists = np.zeros(n_labels)
+    centroids = np.zeros((n_labels, len(X[0])), np.float32)
```
I don't understand why you use the `np.float32`. Maybe, can you adapt to the type of X as it is done in [l174](https://github.com/scikit-learn/scikit-learn/pull/7942/files#diff-1190a10b67dd12abc6c869b9ba329f8aR174) ?

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+----------------------
+
+If the ground truth labels are not known, the Davies–Bouldin index
+(:func:`sklearn.metrics.davies_bouldin_index`) can be used to evaluate the
```
You need an entry in classes.rst for this to work. You should have an entry there anyway!

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1569,3 +1569,77 @@ Drawbacks
+  D_{ij} = \frac{\bar{d_i}+\bar{d_j}}{d_ij}
+
+:math:`\bar{d_i}` is the average distance between each point in cluster
+:math:`i` and the centroid of cluster :math:`i`.
```
I think, "known as its diameter." Then you may add a comment that dbar_j is similarly the diameter for cluster j, or you can just leave it out, because that's clear from the notation.

jnothman commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -255,3 +255,55 @@ def calinski_harabaz_score(X, labels):
+
+    The index is defined as the ratio of within-cluster
+    and between-cluster distances.
+
```
We usually have a link to the user guide here

jnothman commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -255,3 +255,55 @@ def calinski_harabaz_score(X, labels):
+    centroids = np.zeros((n_labels, len(X[0])), np.float32)
+    for k in range(n_labels):
+        cluster_k = X[labels == k]
+        mean_k = np.mean(cluster_k, axis=0)
```
use `cluster_k.mean(axis=0)` and use `safe_indexing` on the line before, and then I think your code will automatically work for sparse matrices...? If there is no reason not to support a sparse matrix input, please add a test and update the docstring.

jnothman commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -255,3 +255,55 @@ def calinski_harabaz_score(X, labels):
+        intra_dists[k] = np.average(pairwise_distances(cluster_k, [mean_k]))
+    centroid_distances = pairwise_distances(centroids)
+    with np.errstate(divide='ignore', invalid='ignore'):
+        if np.all((intra_dists[:, None] + intra_dists) == 0.0) or \
```
why don't you just test `if not np.any(intra_dists)` (or `if np.allclose(intra_dists, 0)` if numerical stability is a concern)?

jnothman commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -255,3 +255,55 @@ def calinski_harabaz_score(X, labels):
+        if np.all((intra_dists[:, None] + intra_dists) == 0.0) or \
+           np.all(centroid_distances == 0.0):
+            return 0.0
+        scores = (intra_dists[:, None] + intra_dists)/centroid_distances
```
space around `/`, please

jnothman reviewed (changes requested):
This more-or-less LGTM!

raghavrv commented:
@tomron Are you with us? :)
